File: Core/Models/Character.cs
using System;
using System.Collections.Generic;
using System.Text;

using Dapper;
using MySql.Data.MySqlClient;

using FW.Core.DbModels;

namespace FW.Core.Models
{
	public class Character
	{
		#region Attribute Properties

		public Dictionary<string, int> Attributes { get; set; }
		public Classes                 Class { get; set; }
		public double                  Drunk { get; set; }
		public double                  Fatigue { get; set; }
		public double                  Luck { get; set; }
		public double                  Mental { get; set; }
		public Poses                   Pose { get; set; }
		public Races                   Race { get; set; }

		#endregion

		#region Connection Properties

		public ConnectionStates        ConnectionState { get; set; }
		public DateTime                Connected { get; set; }
		public DateTime                Created { get; set; }
		public TimeSpan                PlayedTime { get; set; }
		public string                  Prompt { get; set; }
		public bool                    ShowColor { get; set; }
		public int                     SocketID { get; set; }

		#endregion

		#region Ident Properties

		public Alivenesses             Aliveness { get; set; }
		public DateTime                Birthdate { get; set; }
		public Citizenships            Citizenship { get; set; }
		public Location                Location { get; set; }
		public Mortalities             Mortality { get; set; }
		public string                  Name { get; set; }
		public int                     Vnum { get; set; }

		#endregion


		public Character()
		{
			this.Attributes = new Dictionary<string, int>();

			return;
		}


		public void HydrateFromPlayerChar(PlayerChar Char)
		{
			this.Class       = (Classes)Char.Class;
			this.Drunk       = Char.Drunk;
			this.Fatigue     = Char.Fatigue;
			this.Luck        = Char.Luck;
			this.Mental      = Char.Mental;
			this.Pose        = (Poses)Char.Pose;
			this.Race        = (Races)Char.Race;
			this.Created     = Char.Created ?? DateTime.UtcNow;
			this.PlayedTime  = TimeSpan.FromMilliseconds(Char.PlayedTime);
			this.Prompt      = Char.Prompt;
			this.Aliveness   = (Alivenesses)Char.Aliveness;
			this.Birthdate   = Char.Birthdate ?? DateTime.UnixEpoch;
			this.Citizenship = (Citizenships)Char.Citizenship;
			this.Location    = new Location(new System.Numerics.Vector3(Char.PosX, Char.PosY, Char.PosZ), Char.PosVnum);
			this.Mortality   = (Mortalities)Char.Mortality;
			this.Name        = Char.Name;
			this.Vnum        = Char.Vnum;

			return;
		}

		public void SaveToDb(MySqlConnection Conn, bool AddPlayedTime = false)
		{
			StringBuilder sql = new("UPDATE `PlayerChar` SET ");
			sql.Append("`Class` = @Class, ");
			sql.Append("`Drunk` = @Drunk, ");
			sql.Append("`Fatigue` = @Fatigue, ");
			sql.Append("`Luck` = @Luck, ");
			sql.Append("`Mental` = @Mental, ");
			sql.Append("`Pose` = @Pose, ");
			sql.Append("`Race` = @Race, ");
			sql.Append("`Prompt` = @Prompt, ");
			sql.Append("`Aliveness` = @Aliveness, ");
			sql.Append("`Birthdate` = @Birthdate, ");
			sql.Append("`Citizenship` = @Citizenship, ");
			sql.Append("`Mortality` = @Mortality, ");
			sql.Append("`Name` = @Name, ");
			sql.Append("`NameLowered` = @NameLowered, ");
			sql.Append("`PosX` = @PosX, ");
			sql.Append("`PosY` = @PosY, ");
			sql.Append("`PosZ` = @PosZ, ");
			sql.Append("`PosVnum` = @PosVnum");

			if (AddPlayedTime) {
				this.PlayedTime += (DateTime.UtcNow - this.Connected);
				this.Connected   = DateTime.UtcNow;

				sql.Append($", `PlayedTime` = {this.PlayedTime.TotalMilliseconds}");
			}

			sql.Append(" WHERE `Vnum` = @Vnum LIMIT 1");

			Conn.Execute(sql.ToString(), new {
				this.Class,
				this.Drunk,
				this.Fatigue,
				this.Luck,
				this.Mental,
				this.Pose,
				this.Race,
				this.Prompt,
				this.Aliveness,
				this.Birthdate,
				this.Citizenship,
				this.Mortality,
				this.Name,
				NameLowered = this.Name.ToLower(),
				PosX        = this.Location.Coordinate.X,
				PosY        = this.Location.Coordinate.Y,
				PosZ        = this.Location.Coordinate.Z,
				PosVnum     = this.Location.Vnum,
				this.Vnum
			});

			return;
		}
	}
}

File: src/fw/model/character.rs
use crate::fw::model::{ enums, location };

use chrono::{DateTime, Utc};
use std::collections::HashMap;
use std::fmt;
use std::sync::Arc;
use std::time::Duration;
use crate::{SafeQueue, SockMsg};

pub struct Character {
    // Attribute properties
    pub attributes: HashMap<String, u32>,
    pub class: enums::Classes,
    pub drunk: f32,
    pub fatigue: f32,
    pub luck: f32,
    pub mental: f32,
    pub pose: enums::Poses,
    pub race: enums::Races,
    // Connection properties
    pub connection_state: enums::ConnectionStates,
    pub created: DateTime::<Utc>,
    pub played_time: Duration,
    pub prompt: String,
    pub show_color: bool,
    pub socket_id: i32,
    // Ident properties
    pub aliveness: enums::Alivenesses,
    pub birthdate: DateTime::<Utc>,
    pub citizenship: enums::Citizenships,
    pub location: location::Location,
    pub mortality: enums::Mortalities,
    pub name: String,
    pub vnum: u32,
    pub cmd_queue: Vec<String>,
    pub chan_send: Arc<SafeQueue<SockMsg>>,
    pub chan_recv: Arc<SafeQueue<SockMsg>>,
}

impl fmt::Display for Character {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Char '{}' #{}", self.name, self.vnum)
    }
}

impl Character {
    pub fn new() -> Character {
        Character {
            attributes: HashMap::new(),
            class: enums::Classes::None,
            drunk: 0.0,
            fatigue: 0.0,
            luck: 0.0,
            mental: 0.0,
            pose: enums::Poses::Standing,
            race: enums::Races::Ebban,
            connection_state: enums::ConnectionStates::Disconnected,
            created: Utc::now(),
            played_time: Duration::new(0, 0),
            prompt: String::from(""),
            show_color: false,
            socket_id: 0,
            aliveness: enums::Alivenesses::Dead,
            birthdate: Utc::now(),
            citizenship: enums::Citizenships::None,
            location: location::Location::empty(),
            mortality: enums::Mortalities::Mortal,
            name: String::from(""),
            vnum: 0,
            cmd_queue: Vec::new(),
            chan_send: Arc::new(SafeQueue::<SockMsg>::new()),
            chan_recv: Arc::new(SafeQueue::<SockMsg>::new()),
        }
    }

    pub fn new_from_vnum(vnum: u32) -> Character {
        let mut tmp = Character::new();
        tmp.vnum = vnum;

        tmp
    }
}
